### ISSUE TITLE
np.sqrt encounters negative values in an array of zeros

### ISSUE THREAD

**Issue by jdavies-st** [ContinuumIO/anaconda-issues#9129]
For a large enough array of zeros, `np.sqrt` throws a warning indicating that it encountered a negative value in that array of zeros:
```python
In [1]: import numpy as np

In [2]: a = np.zeros((100, 100), dtype=np.float32)

In [3]: np.sqrt(a)
/Users/jdavies/anaconda/envs/jwst_devel/bin/ipython:1: RuntimeWarning: invalid value encountered in sqrt
  #!/Users/jdavies/anaconda/envs/jwst_devel/bin/python
Out[3]: 
array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)
```
My expectation would be that the square root of zeros would also produce zeros, with no negative values.

This is not happening on Linux - only on MacOS (tested on versions 10.12.6 and 10.13.4). It only happens when the array is large enough and `dtype=np.float32`.  For a small 3x3 array, the zeros are all zeros.  Doing `np.sqrt(np.abs(a))` throws the same warning.

#### Operating System:

MacOS (tested on versions 10.12.6 and 10.13.4)

#### conda info

```
     active environment : jwst_devel
    active env location : /Users/jdavies/anaconda/envs/jwst_devel
            shell level : 1
       user config file : /Users/jdavies/.condarc
 populated config files : /Users/jdavies/.condarc
          conda version : 4.5.0
    conda-build version : 1.18.2
         python version : 3.5.1.final.0
       base environment : /Users/jdavies/anaconda  (writable)
           channel URLs : http://ssb.stsci.edu/astroconda/osx-64
                          http://ssb.stsci.edu/astroconda/noarch
                          https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/osx-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/osx-64
                          https://repo.anaconda.com/pkgs/pro/noarch
          package cache : /Users/jdavies/anaconda/pkgs
                          /Users/jdavies/.conda/pkgs
       envs directories : /Users/jdavies/anaconda/envs
                          /Users/jdavies/.conda/envs
               platform : osx-64
             user-agent : conda/4.5.0 requests/2.14.2 CPython/3.5.1 Darwin/16.7.0 OSX/10.12.6
                UID:GID : 1701:1031
             netrc file : None
           offline mode : False
```

#### `conda list --show-channel-urls`

[conda_list_show_channel_urls.txt](https://github.com/ContinuumIO/anaconda-issues/files/1891479/conda_list_show_channel_urls.txt)

[conda_list_explicit.txt](https://github.com/ContinuumIO/anaconda-issues/files/1891481/conda_list_explicit.txt)

**Comment by jdavies-st** [comment]
A bit of further information.  The numpy version that gives me the RuntimeWarning is:
```python
In [21]: np.__version__
Out[21]: '1.14.2'

In [22]: np.__config__.show()
mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/include']
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/jdavies/anaconda/envs/jwst_devel/include']
```
whereas running the same code above in an old conda environment does not give the RuntimeWarning.  Old conda environment below:
```python
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12) 
Type "copyright", "credits" or "license" for more information.

In [1]: import numpy as np

In [2]: a = np.zeros((100, 100), dtype=np.float32)

In [3]: np.sqrt(a)
Out[3]: 
array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]], dtype=float32)

In [4]: np.__version__
Out[4]: '1.11.2'

In [5]: np.__config__.show()
blas_opt_info:
    include_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/include']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    library_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/lib']
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
blas_mkl_info:
    include_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/include']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    library_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/lib']
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
openblas_lapack_info:
  NOT AVAILABLE
lapack_opt_info:
    include_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/include']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    library_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/lib']
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
lapack_mkl_info:
    include_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/include']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    library_dirs = ['/Users/jdavies/anaconda/envs/jwstb7/lib']
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
```

**Comment by jjhelmus** [comment]
This seems to be a side effect of the NumPy 1.14.2 build 1 which included some patches to make better use of MKL, build 0 does not exhibit this behavior.  @stuartarchibald can you take a look into this?

**Comment by stuartarchibald** [comment]
An explanation of what I think might be going on...

The patches in build 1 make use of Intel's VML (vector math library) for performing vectorized math operations as needed by NumPy `ufuncs`. As VML is a library, function calls have to be made to use it and this, along with a few other things, cause some minor overheads. The net result is that for "small" vectors it's cheaper to just call an inlined unary loop around the `math.h` equivalent function and for "large" vectors the cost of making the function call etc is amortized in the higher performance of the VML implementation of the function. In practice different groups of functions have different thresholds for "small" vs. "large" behaviour, `sqrt` is under the transcendental group and VML is used for vectors of length `>8192`, hence `100*100` trips it. I think this explains the "it only happens when the array is large enough" statement.

To the rest of the explanation, I don't think NumPy is raising that warning because it has encountered a negative number on input. I think it is raising that warning because it is checking the floating point exception flags on return from the `ufunc` calling into compiled code and it so happens a `FE_INVALID` is present from the call to VML's single precision vector sqrt function (`vsSqrt()`). NumPy therefore "thinks" that there has been a domain error and so raises the warning.

This C code replicates:
```C
#include <fenv.h>                                                                   
#include <string.h>                                                                 
#include <stdio.h>                                                                  
#include <stdlib.h>                                                                 
#include "mkl.h"                                                                    
                                                                                    
#define TYPE float                                                                  
#define FUNC vsSqrt                                                                 
                                                                                    
#pragma STDC FENV_ACCESS ON                                                         
                                                                                    
// From: http://en.cppreference.com/w/c/numeric/fenv/feexceptflag       
// For demonstration purposes only.            
void show_fe_exceptions(void)                                                       
{                                                                                   
    printf("current exceptions raised: ");                                          
    if(fetestexcept(FE_DIVBYZERO))     printf(" FE_DIVBYZERO");                     
    if(fetestexcept(FE_INEXACT))       printf(" FE_INEXACT");                       
    if(fetestexcept(FE_INVALID))       printf(" FE_INVALID");                       
    if(fetestexcept(FE_OVERFLOW))      printf(" FE_OVERFLOW");                      
    if(fetestexcept(FE_UNDERFLOW))     printf(" FE_UNDERFLOW");                     
    if(fetestexcept(FE_ALL_EXCEPT)==0) printf(" none");                             
    printf("\n");                                                                   
}                                                                                   
                                                                                    
int main(void) {                                                                    
                                                                                    
TYPE * a, * b;                                                                      
int n = 100;                                                                        
int err;                                                                                                                                            
                                                                                    
a = (TYPE *)malloc(n * n * sizeof(TYPE));                                           
b = (TYPE *)malloc(n * n * sizeof(TYPE));                                           
                                                                                    
for (size_t i = 0; i < n; i++) {                                                    
    a[i] = 0.;                                                                         
}                                                                                   
                                                                                    
printf("pre call:");                                                                
show_fe_exceptions();                                                               
FUNC(n, a, b);                                                                      
printf("post call:");                                                               
show_fe_exceptions();                                                               
err = vmlGetErrStatus();                                                            
printf("errno %s\n", strerror(errno));                                              
printf("%d\n", err);                                                                
                                                                                    
                                                                                    
free(a);                                                                            
free(b);                                                                            
return 0;                                                                           
                                                                                    
}; 
```
gives:
```
pre call:current exceptions raised:  none
post call:current exceptions raised:  FE_INEXACT FE_INVALID
errno No such file or directory
0
```
Whereas if the `TYPE=double` and the `FUNC=vdSqrt`, the output is:
```
pre call:current exceptions raised:  none
post call:current exceptions raised:  FE_INEXACT
errno No such file or directory
0
```
I suspect the fix is going to be to clear all `FE_` flags post VML call, inspect the VML status and if the VML status has a floating point exception set (probably `VML_STATUS_ERRDOM|VML_STATUS_SING|VML_STATUS_UNDERFLOW|VML_STATUS_OVERFLOW`) then set the equivalent `FE_` flag.

CC @oleksandr-pavlyk

**Comment by oleksandr-pavlyk** [comment]
@stuartarchibald Yes, I agree. I will also make sure to file an issue with MKL team.

**Comment by jdavies-st** [comment]
Thanks for looking into this.  I verify that on my system, build 0 of 1.14.2 does not exhibit this behavior.

In the meantime, is there a way as a user to turn off mkl acceleration for Anaconda-built numpy?  I notice when I do a `conda search numpy` that there are `nomkl` versions
```
numpy                    1.14.2 py36_nomklh9797aa9_0  pkgs/main           
numpy                    1.14.2 py36_nomklh9797aa9_1  pkgs/main           
numpy                    1.14.2  py36ha9ae307_0  pkgs/main           
numpy                    1.14.2  py36ha9ae307_1  pkgs/main           
```
but I can't get them to resolve
```
(testnomkl) solo> conda install numpy=1.14.2=py36_nomklh9797aa9_1
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - numpy==1.14.2=py36_nomklh9797aa9_1
  - *[track_features=nomkl]

Current channels:

  - http://ssb.stsci.edu/astroconda/osx-64
  - http://ssb.stsci.edu/astroconda/noarch
  - https://repo.anaconda.com/pkgs/main/osx-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/free/osx-64
  - https://repo.anaconda.com/pkgs/free/noarch
  - https://repo.anaconda.com/pkgs/r/osx-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/pro/osx-64
  - https://repo.anaconda.com/pkgs/pro/noarch
```
I must be doing something wrong, but I can't find any documentation of how a user can install the `nomkl` feature.

**Comment by jjhelmus** [comment]
You should be able to install the version of NumPy which uses OpenBLAS using `conda install numpy nomkl`.  The `nomkl` package is the key install the `nomkl` feature.

**Comment by jdavies-st** [comment]
That does it.  Many thanks.  I understand feature tracking a bit better now.  😎

**Comment by slinnarsson** [comment]
**tl;dr**: Pure numpy `np.sqrt()` seems to be as fast or even faster than Anaconda's MKL-accelerated version, so it would likely be better to just leave it alone.

So, over at the numpy repo, @juliantaylor claims that the built-in sqrt is already fast. If that's true, a better fix would be to not call MKL at all, rather than try to work around the bug it causes. To test this idea, I ran the following test code:

```python
 import time
import numpy as np
x = []
y = []
for n in range(8,24):
    a = np.zeros(2**n).astype('float32')
    temp = []
    before = time.time()
    for i in range(500):
        _ = np.sqrt(a)
    after = time.time()
    temp.append(after - before)
    x.append(n)
    y.append(np.mean(temp) / 2**n / 500)
print(x)
print(y)
```

Essentially, just computing `np.sqrt(a)` for all-zero vectors of length 2^8 through 2^24, and averaging over 500 runs. 

With Anaconda numpy using MKL (blue), execution is consistently slightly slower than using pure numpy (red), at all vector sizes:

![np sqrt](https://user-images.githubusercontent.com/10656387/43369403-868d71bc-9322-11e8-8443-14769fca4b1f.png)

Given that the difference is independent of vector size, it seems that it's MKL itself that's slower. I guess this could depend very much on the exact processor architecture, and numpy just happens to run slightly better on my particular machine. Changing from `float32` to `float64` makes no difference (i.e. numpy is still faster).

My Anaconda Python 3.6 numpy-with-MKL says:

```python
>>> np.show_config()
mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/stelin/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/stelin/anaconda/include']
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/stelin/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/stelin/anaconda/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/stelin/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/stelin/anaconda/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/stelin/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/stelin/anaconda/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/Users/stelin/anaconda/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/stelin/anaconda/include']
```

And my MacPorts Python 3.6 pure numpy says:

```python
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
blas_opt_info:
    extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_3_10_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
atlas_info:
  NOT AVAILABLE
lapack_opt_info:
    extra_compile_args = ['-msse3']
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
```

I was running this on a MacBook Pro 15 inch late 2016 model (2.9 GHz Intel Core i7).

**Comment by msarahan** [comment]
Cc @oleksandr-pavlyk

**Comment by slinnarsson** [comment]
The difference in performance pretty much goes away if you start with uniform random numbers instead of all zeros:

![np sqrt](https://user-images.githubusercontent.com/10656387/43372125-3d416b32-9352-11e8-9843-db92d83c46e7.png)

(i.e. here I'm using `np.random.uniform(size=2**n)` instead of `np.zeros(2**n)`). Maybe MKL gets distracted by all those zeros?

**Comment by oleksandr-pavlyk** [comment]
I would like to point out that the issue with MKL `vdExp` raising `FE_INVALID` flag on MacOSX is fixed MKL 2019. 

The errors are benign in that they do not affect correctness of the result,, but it is annoying, and may cause problems in tests which check that no run-time warnings were issued.

**Comment by jdavies-st** [comment]
I've verified this is fixed and am closing the issue.
```
(jwst_dev) rey> conda list | grep numpy
numpy                     1.15.2           py37h6a91979_1  
numpy-base                1.15.2           py37h8a80b8c_1  
(jwst_dev) rey> conda list | grep mkl
blas                      1.0                         mkl  
mkl                       2019.0                      118  
mkl_fft                   1.0.6            py37hb8a8100_0  
mkl_random                1.0.1            py37h5d10147_1  
(jwst_dev) rey> ipython
Python 3.7.0 (default, Jun 28 2018, 07:39:16) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.5.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: a = np.zeros((100, 100), dtype=np.float32)

In [3]: np.sqrt(a)
Out[3]: 
array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)
```

Many thanks

**Comment by msarahan** [comment]
Thanks for the confirmation, @jdavies-st
